### ISSUE TITLE
Custom handler monads and partial handlers.

### ISSUE THREAD

**Issue by mf59816** [haskell-servant/servant#87]
I am trying to implement authentication and got stuck with errors that
I think could more sustainably be fixed in the 'HasServer' class.
This is a recount of what happened, with a concrete question at the
end.

Relevant code is here: https://github.com/liqd/thentos/commit/93016c1c3d04d1b48ce236ad5a4a720c410f6c9b (branch: liqd/thentos#81).

Very similar things are going on over at #70, but thentos writes the
credentials into the custom handler monad instead of passing them to
the handlers as an extra argument.  My mind is a bit boggled right
now, but I think this is what's going on:

1. At the point in the route where the `ThentosAuth` link is found, a
__"partial server"__ in the custom handler monad with result type `()`
is constructed: Just look up the credentials in the request, write
them into the custom monad's state, and return `()`.

2. This partial server cannot be executed here, because we are still
collecting the handler's function arguments from the routes.  Actions
can be __constructed__ half-way through the route, but they can't be
__executed__ before the entire route has been parsed successfully.

3. So instead, I push the action down the routes to where I can
sequence it into the route's "proper" action.  (@jkarni has showed me
how to implement an earlier version of this trick.)

Now as you can see in module `Thentos.Backend.Api.Auth` if you
uncomment the code in lines 51, 54, `HasServer` wants sub-routes to
live in the standard servant monad.

I guess my question is, what would happen if the 'HasServer' class
looked like this?:

```haskell
class HasServer layout (m :: * -> *) where
  type ServerT layout :: *
  route :: Proxy layout -> ServerT layout -> RoutingApplication
```

Also, is there a work-around?

Thanks!  (-:

**Comment by alpmestan** [comment]
> I guess my question is, what would happen if the 'HasServer' class looked like this?:
>
> class HasServer layout (m :: * -> *) where
>   type ServerT layout :: *
>   route :: Proxy layout -> ServerT layout -> RoutingApplication
>
> Also, is there a work-around?

If `HasServer` was defined this way, how could GHC guess which `m` you want to use when looking at which `HasServer` instances it'll pick?

**Comment by mf59816** [comment]
Right.

Would it help if I added `m` as a parameter to `ServerT` again?

```haskell
class HasServer layout (m :: * -> *) where
  type ServerT layout m :: *
  route :: Proxy layout -> ServerT layout m -> RoutingApplication
```

I think I originally added `m` to the class parameters to bind it in the common scope of `ServerT` and `route`, and then thought I could drop it from `ServerT`.  Dumm-dumm.

Anyway, I should think this through some more.  If you invite me to this repo, I can label my issues "low prio" in the future.  :-)

**Comment by alpmestan** [comment]
I see two problems with this approach (but would love being proven wrong).

- by non-injectivity of type families, we can't "reverse-compute" `m` based on `ServerT layout m` so there still is an instance selection problem here ;
- `m` kind of "isn't meant" to be an argument to `HasServer`. You might wonder why? Because `m` only affects the (return) types of the handlers, everything else stays the same even when you change the monad. Whereas here, you'd have to either do one of:

    ``` haskell
    --  forall m :: * -> *
    instance (...) => HasServer (Get cts a) m where .....

    -- for specific m's...
    instance (...) => HasServer (Get cts a) (EitherT ServantErr IO) where ...
    instance (...) => HasServer (Get cts a) (Reader String) where ...
    ```

In the first case, you are basically saying you can provide a way to go from `m a` to `RoutingApplication` for any `m`, which is not possible, we need to know more about `m`.

In the second case, that simply won't scale. Instead, `servant-0.4` comes with a little machinery to use other monads (or types of kind `* -> *` more generally) as explained [here](http://haskell-servant.github.io/tutorial/server.html#using-another-monad-for-your-handlers). This lets you just provide the minimal amount of boilerplate (you say just once how to go from your custom handler monad to `EitherT ServantErr IO`).

Don't worry about priorities etc, it's always nice to have people poke at the code and research improvements to it =)

**Comment by mf59816** [comment]
Ok, that is very helpful.  Let's see.

- Non-injectivity is a technical problem.  there may be work-arounds or even solid solutions, but your second point is much stronger.

- Of course I don't want to write monad-specific instances.  And I also don't something that doesn't make sense at all.  (-:

So somebody else will have to prove you wong, meanwhile I withdraw my proposal :-).  Appearently the solution has to look quite different.

I'm quite excited about `enter`, but it's not enough for what I am trying to do.  Let me try to explain again.

Informally (and possibly wrongly), let's call handler functions that work on routes including the verb type **saturated** and handlers that work on prefixes of such routes **partial**.

An application for partial handlers is suggested by #70: Instead of passing along the credentials collected when processing the `BasicAuth` segment of the path and having to take them as extra arguments in all handlers, my plan is to change the state of my custom handler monad here, resume processing of the route without having generated an extra handler argument, and allow the handlers to remain completely oblivious of what just happened.

Sligthly off-topic: if the custom monad is a stack based in [lio](http://hackage.haskell.org/package/lio), the handlers not only will have no idea that they run with limited privileges, but they will be unable to escalate those privileges even if they want to.  This makes it possible to e.g. write a dialect of a protocol implementation that has been proven secure, and have the same secrecy and integrity guarantees in your code as in the original implementation, even if you deliberately write malicious handlers.

So that's what I'm after.  How do I get it?  I'll be back after more thinking.

**Comment by alpmestan** [comment]
Very interesting direction. Feel free to just keep this issue open and use it as you please (reporting progress/ideas and what not).

**Comment by mf59816** [comment]
Hum, @jkarni took a look off-line, and suddenly it wasn't that hard: https://github.com/liqd/thentos/commit/be1800607282bcc57e4ed37f9fef984baca67645.

It works fine now since thentos has only one point at the top of the api where authentication is performed.  Open problems:

1. I don't know what I could do if I wanted to `enter` the monad several times during processing of the route.  That's what I originally tried, and it was too hard for me.  Now I don't have an application to a possible solution, so the problem is not that urgent for me.  (-:
2. I would like to have a function `thentosAuth` that lets me `enter` more conveniently into the `Action` monad, but I can't because of non-injectivity of type families.  See haddock comment in `Thentos.Backend.Simple.Auth` in the above commit.

I'll close this issue for now, since I'm not sure how soon I'll get back to it, and you have enough stuff going on here for people to stay entertained.  (-:

Thanks for listening everybody, and thanks @alpmestan and @jkarni for the great support!
